prov/efa/src/rxr/rxr_ep.c

```diff
@@ -1637,26 +1644,43 @@ void rxr_ep_progress_internal(struct rxr_ep *ep)
 	dlist_foreach_container_safe(&ep->tx_entry_queued_list,
 				     struct rxr_tx_entry,
 				     tx_entry, queued_entry, tmp) {
-		if (tx_entry->state == RXR_TX_QUEUED_CTRL)
-			ret = rxr_pkt_post_ctrl(ep, RXR_TX_ENTRY, tx_entry,
-						tx_entry->queued_ctrl.type,
-						tx_entry->queued_ctrl.inject);
-		else
-			ret = rxr_ep_send_queued_pkts(ep, &tx_entry->queued_pkts);
-
+		/*
+		 * It is possible to receive an RNR after we queue this
+		 * tx_entry if we run out of resources in the medium message
+		 * protocol. Ensure all queued packets are posted before
+		 * continuing to post additional control messages.
+		 */
+		ret = rxr_ep_send_queued_pkts(ep, &tx_entry->queued_pkts);
```

Comment:
My understanding is this mainly affect medium message protocol, which will send multiple MEDIUM_RTM packets, each packet is a REQ packet.

As a result, tx_entry can be in "queued_ctrl" state (outstanding TX operation reached max) and "queued_rnr" state (MEDIUM_RTM already sent encounter RNR).

With the change applied, could the following scenario still happen?

1. tx_entry send multiple MEDIUM_RTM packets,
2. first several was sent, then it encountered -FI_EAGAIN.
3. tx_entry was put in ```tx_entry_quened_list```, state was set to TX_QUEUED_CTRL.
4. a MEDIUM_RTM packet sent encountered RNR, tx_entry state was set to TX_QUEUED_REQ_RNR.
5. entering ep_progress_internal(), we first send RNR packets.
6. After RNR packets were sent, because the state is not RXR_TX_QUEUED_CTRL, we did not continue to send the rest of the MEDIUM_RTM packets.

Reply:
Yes, that's right that this mainly affects medium message as it posts multiple REQ packets. The main issue was not setting the state back to `RXR_TX_REQ` on L1682 which caused TX entries to be removed from the queued list before all packets are sent.

The scenario you mentioned will not happen. However, it's not really intuitive and I wasn't sure how to improve the case you mentioned without more changes to our state machine. We will never transition from `TX_QUEUED_CTRL` to `TX_QUEUED_REQ_RNR` (see `rxr_cq_handle_cq_error()`). We can do that since those both of those states are handled in the loop here and the entries remain on the same list.

Three scenarios are possible here with medium message:
- Peer is in backoff, return -FI_EAGAIN from fi_send
- All messages are posted, one or more gets RNR and transitions to TX_QUEUED_REQ_RNR. Backoff/retry happens for those packets.
- Some messages are posted but resources are exhausted and transitions to TX_QUEUED_CTRL. If the previously posted messages get RNR we stay in TX_QUEUED_CTRL and send the queued RNR packets before trying to post the remaining packets.

Reply:
I see. I read the ```rxr_cq_handle_cq_error``` code again and understand what you mean now. We will not transit from ```QUEUE_CTRL``` to ```QUEUED_REQ_RNR```.

Reply:
Actually, making this clearer was easier than I thought. I wrote a patch [here](https://github.com/ofiwg/libfabric/commit/0381c038495e40f1a02afc734bde7d93c382bc3f) if you want to take a look that I'll test later.

Reply:
Thanks! I like the patch as it make the code much more clear. Looking forward to your new PR!